testing/mock/mock_opae.h

```diff
@@ -0,0 +1,77 @@
+// Copyright(c) 2017-2019, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+/*
+ * mock_opae.h
+ */
+#pragma once
+
+
+
+#include <opae/fpga.h>
+#include "gtest/gtest.h"
+#include "test_system.h"
+
+namespace opae {
+namespace testing {
+
+template<int _T = 2>
+class mock_opae_p : public ::testing::TestWithParam<std::string> {
+ protected:
+  mock_opae_p(): tokens_{ {} } {}
+
+  virtual void SetUp() override {
+    ASSERT_TRUE(test_platform::exists(GetParam()));
+    platform_ = test_platform::get(GetParam());
+    system_ = test_system::instance();
+    system_->initialize();
+    system_->prepare_syfs(platform_);
+    test_setup();
+  }
+
+  virtual void TearDown() override {
+    for (auto &t : tokens_) {
+      if (t) {
+        EXPECT_EQ(fpgaDestroyToken(&t), FPGA_OK);
+        t = nullptr;
+      }
+    }
+    test_teardown();
+    system_->finalize();
+  }
+
+  virtual void test_setup() {
```

Comment:
Just curious.. is there an advantage of providing these virtual overrides vs using the SetUp and TearDown that are already provided by gtest?  (A derived class could be required to call the parent implementation.)

Reply:
The SetUp and TearDown provided by gtest are the interfaces for gtest meaning they will always be called by gtest. The test_setup/test_teardown are the opae mock interface. The mock base class implements just enough boiler plate in the SetUp/TearDown and then calls test_setup/test_teardown which can be overridden by derived classes. This means the derived classes don't have to worry about mock setup (test_system/mock sysfs) and only concentrate on their specific routines.